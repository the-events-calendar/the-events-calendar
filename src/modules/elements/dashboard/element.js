/**
 * External Dependencies
 */
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 * Internal dependencies
 */
import './style.pcss';

/**
 * Enumeration with the available directions.
 *
 * @type {{up: string, down: string}}
 */
export const directions = {
	up: 'up',
	down: 'down',
};

/**
 * Usage of this component:
 *
 * <Dashboard isOpen={true} className="custom" direction={directions.up}>
 *   <AnyComponent></AnyComponent>
 * </Dashboard>
 */

const Dashboard = ( {
	className = '',
	direction = directions.down,
	isOpen = false,
	children = null,
} ) => {
	const containerClasses = classNames(
		'tribe-editor__dashboard__container',
		`tribe-editor__dashboard__container--${ direction }`,
		{ 'tribe-editor__dashboard__container--open': isOpen },
		className,
	);

	return (
		<div className={ containerClasses }>
			<div className="tribe-editor__dashboard">
				{ children }
			</div>
		</div>
	);
};

Dashboard.propTypes = {
	isOpen: PropTypes.bool,
	className: PropTypes.string,
	direction: PropTypes.oneOf( Object.keys( directions ) ),
	children: PropTypes.element,
};

export default Dashboard;
