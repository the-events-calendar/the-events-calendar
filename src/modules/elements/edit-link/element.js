/**
 * External dependencies
 */
import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { globals } from '@moderntribe/common/utils';
import './style.pcss';

const EditLink = ( {
	postId = 0,
	label = __( 'Edit', 'the-events-calendar' ),
	target = '_blank',
} ) => {
	const admin = get( globals.common(), 'adminUrl', '' );
	if ( ! admin || ! postId ) {
		return null;
	}

	const extraProps = {
		rel: '_blank' === target ? 'noreferrer noopener' : undefined,
	};

	return (
		<a
			className="tribe-editor__edit-link"
			href={ `${ admin }post.php?post=${ postId }&action=edit` }
			target={ target }
			{ ...extraProps }
		>
			{ label }
		</a>
	);
};

EditLink.propTypes = {
	postId: PropTypes.number,
	label: PropTypes.string,
	target: PropTypes.string,
};

export default EditLink;
