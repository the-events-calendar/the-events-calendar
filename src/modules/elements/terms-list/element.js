/**
 * External dependencies
 */
import React from 'react';
import { compose } from 'redux';
import { unescape } from 'lodash';
import { PropTypes } from 'prop-types';

/**
 * WordPress dependencies
 */
import { Spinner } from '@wordpress/components';
import { withSelect } from '@wordpress/data';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import './style.pcss';

const getTerms = ( terms, parentId = null ) => {
	if ( ! terms || ! terms.length ) {
		return [];
	}

	if ( parentId === null ) {
		return terms;
	}

	return terms.filter( ( term ) => term.parent === parentId );
};

const getTermListClassName = ( level = 0 ) => (
	`tribe-editor__terms__list tribe-editor__terms__list--level-${ level }`
);

const getTermListItemClassName = ( level = 0 ) => (
	`tribe-editor__terms__list-item tribe-editor__terms__list-item--level-${ level }`
);

const termName = ( term = {} ) => {
	return term.name
		? unescape( term.name ).trim()
		: __( '(Untitled)', 'the-events-calendar' );
};

const Label = ( { text } ) => (
	<strong className="tribe-editor__terms__label" key="terms-label">
		{ text }
		{ ' ' }
	</strong>
);

Label.propTypes = {
	text: PropTypes.string,
};

const Empty = ( { renderEmpty = null, id, label } ) => (
	renderEmpty && (
		<div key={ id } className="tribe-editor__terms--empty">
			<Label text={ label } />
			{ renderEmpty }
		</div>
	)
);

Empty.propTypes = {
	renderEmpty: PropTypes.node,
	id: PropTypes.string,
	label: PropTypes.string,
};

const List = ( {
	terms = [],
	termSeparator = ', ',
	isLoading = false,
	id = '',
	className = '',
} ) => {
	if ( isLoading ) {
		return <Loading id={ id } className={ className } />;
	}

	return (
		<ul className={ getTermListClassName() }>
			{ terms.map( ( term, index ) => (
				<Item
					key={ index }
					term={ term }
					separator={ termSeparator }
					isLast={ index + 1 === terms.length }
				/>
			) ) }
		</ul>
	);
};

List.propTypes = {
	terms: PropTypes.array,
	termSeparator: PropTypes.string,
	isLoading: PropTypes.bool,
	id: PropTypes.string,
	className: PropTypes.string,
};

const Separator = ( { delimiter, isLast } ) => ! isLast ? <span>{ delimiter }</span> : '';

Separator.propTypes = {
	delimiter: PropTypes.string,
	isLast: PropTypes.bool,
};

const Item = ( { separator, term, isLast } ) => {
	let termLink = term.link;

	// Modifies the tag slug for the post_tag taxonomy to include an "events" prefix.
	if ( 'post_tag' === term.taxonomy ) {
		termLink = '/events/tag/' + term.slug;
	}

	return (
		<li key={ term.id } className={ getTermListItemClassName( 0 ) }>
			<a
				href={ termLink }
				target="_blank"
				rel="noopener noreferrer"
				className="tribe-editor__terms__list-item-link"
			>
				{ termName( term ) }
			</a>
			<Separator delimiter={ separator } isLast={ isLast } />
		</li>
	);
};

Item.propTypes = {
	separator: PropTypes.string,
	term: PropTypes.object,
	isLast: PropTypes.bool,
};

const Loading = ( { id = '', className = '' } ) => (
	<div key={ id } className={ `tribe-editor__terms__spinner ${ className }` }>
		<Label />
		<Spinner key="terms-spinner" />
	</div>
);

Loading.propTypes = {
	id: PropTypes.string,
	className: PropTypes.string,
};

export const TaxonomiesElement = ( {
	className,
	slug,
	label,
	renderEmpty,
	isRequesting,
	...rest
} ) => {
	const terms = getTerms( rest.terms );
	const key = `tribe-terms-${ slug }`;

	if ( ! terms.length && ! isRequesting ) {
		return <Empty id={ key } renderEmpty={ renderEmpty } label={ label } />;
	}

	return (
		<div key={ key } className={ `tribe-editor__terms ${ className }` }>
			<Label text={ label } />
			<div key="terms" className="tribe-editor__terms__list-wrapper">
				<List terms={ terms } className={ className } id={ key } isLoading={ isRequesting } />
			</div>
		</div>
	);
};

TaxonomiesElement.propTypes = {
	className: PropTypes.string,
	slug: PropTypes.string,
	label: PropTypes.string,
	renderEmpty: PropTypes.node,
	isRequesting: PropTypes.bool,
	terms: PropTypes.array,
};

TaxonomiesElement.defaultProps = {
	termSeparator: __( ', ', 'the-events-calendar' ),
	className: '',
	terms: [],
	isRequesting: false,
};

const applySelect = withSelect( ( select, props ) => {
	const { getEntityRecords } = select( 'core' );
	const { isResolving } = select( 'core/data' );
	const { slug } = props;
	// post_tags are stored as 'tags' in the editor attributes
	const attributeName = slug === 'post_tag' ? 'tags' : slug;
	const ids = select( 'core/editor' ).getEditedPostAttribute( attributeName );

	if ( ! ids || ids.length === 0 ) {
		return { terms: [], isRequesting: false };
	}

	const query = {
		orderby: 'count',
		order: 'desc',
		include: ids,
	};

	return {
		terms: getEntityRecords( 'taxonomy', slug, query ),
		isRequesting: isResolving( 'core', 'getEntityRecords', [ 'taxonomy', slug, query ] ),
	};
} );

export default compose(
	applySelect,
)( TaxonomiesElement );
