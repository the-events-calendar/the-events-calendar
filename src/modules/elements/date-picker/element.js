/**
 * External dependencies
 */
import React from 'react';
import moment from 'moment';
import { noop } from 'lodash';
import { PropTypes } from 'prop-types';

/**
 * WordPress dependencies
 */
import { Component } from '@wordpress/element';
import {
	DatePicker as WPDatePicker,
	Dropdown,
} from '@wordpress/components';

/**
 * Internal dependencies
 */
import { moment as momentUtil } from '@moderntribe/common/utils';
import './style.pcss';

const { toMoment, toDate, toDatePicker, toDateTime } = momentUtil;

export default class DatePicker extends Component {
	static propTypes = {
		changeDatetime: PropTypes.func,
		datetime      : PropTypes.string,
	};

	constructor( props ) {
		super( props );

		this.changeDatetime = props.changeDatetime.bind( this );

		this.state = {
			...props,
			datetime: toDatePicker( toMoment( props.datetime ) ),
		};
	}

	static getDerivedStateFromProps( nextProps, prevState ) {
		const { datetime } = nextProps;
		if ( datetime === prevState.datetime ) {
			return null;
		}

		return {
			datetime: toDatePicker( toMoment( datetime ) ),
		};
	}

	normalize = ( date = moment() ) => {
		// Convert to moment
		const current = moment( date );
		return current.isValid() ? current : moment();
	};

	renderContent = ( { onClose } ) => {
		this.onClose = onClose.bind( this );
		const { datetime } = this.state;

		return (
			<WPDatePicker
				key="date-picker"
				currentDate={ moment( datetime ) }
				onChange={ this.onChange }
			/>
		);
	};

	renderToggle = ( { onToggle, isOpen } ) => {
		const { datetime } = this.state;

		return (
			<button
				type="button"
				className="button-link"
				onClick={ onToggle }
				aria-expanded={ isOpen }
			>
				{ toDate( this.normalize( datetime ) ) }
			</button>
		);
	};

	onChange = ( date ) => {
		this.changeDatetime( toDateTime( this.normalize( date ) ) );
		this.onClose();
	};

	render() {
		return (
			<Dropdown
				className="tribe-editor__datepicker"
				position="bottom left"
				contentClassName="tribe-editor__datepicker-dialog"
				renderToggle={ this.renderToggle }
				renderContent={ this.renderContent }
			/>
		);
	}
}
