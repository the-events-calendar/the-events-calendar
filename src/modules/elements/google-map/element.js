/**
 * External dependencies
 */
import { stringify } from 'querystringify';
import { values, noop, isArray, isEmpty, get } from 'lodash';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

import { Component } from '@wordpress/element';

import {
	Spinner,
} from '@wordpress/components';

/**
 * Internal dependencies
 */
import './style.pcss';
import { google, mapsAPI } from '@moderntribe/common/utils/globals';

/**
 * Module Code
 */
const IMAGE_FORMATS = {
	// png8 or png (default) specifies the 8-bit PNG format.
	PNG: 'png',

	// png32 specifies the 32-bit PNG format.
	PNG32: 'png32',

	// gif specifies the GIF format.
	GIF: 'gif',

	// jpg specifies the JPEG compression format.
	JPG: 'jpg',

	// jpg-baseline specifies a non-progressive JPEG compression format.
	JPG_BASELINE: 'jpg-baseline',
};

const MAP_TYPES = {
	// roadmap (default) specifies a standard roadmap image, as is normally shown on the Google Maps website.
	ROADMAP: 'roadmap',

	// satellite specifies a satellite image.
	SATELLITE: 'satellite',

	// terrain specifies a physical relief map image, showing terrain and vegetation.
	TERRAIN: 'terrain',

	// hybrid specifies a hybrid of the satellite and roadmap image,
	// showing a transparent layer of major streets and place names on the satellite image.
	HYBRID: 'hybrid',
};

const IMAGE_FORMATS_VALUES = values( IMAGE_FORMATS );
const MAP_TYPES_VALUES = values( MAP_TYPES );

const API_KEY = get( mapsAPI(), 'key', '' );
const DEFAULT_ZOOM = parseInt( get( mapsAPI(), 'zoom', 14 ), 10 );

/**
 * A wrapper for Google's Static Maps
 *
 * @see https://developers.google.com/maps/documentation/staticmaps/intro#Overview
 * @example: http://staticmapmaker.com/google/
 */

export default class GoogleMap extends Component {
	static RootStaticUrl = 'https://maps.googleapis.com/maps/api/staticmap';
	static RootEmbedUrl = 'https://www.google.com/maps/embed/v1/place';

	static ImageFormats = IMAGE_FORMATS

	static MapTypes = MAP_TYPES

	static propTypes = {
		coordinates: PropTypes.object.isRequired,
		address: PropTypes.string,
		size: PropTypes.shape( {
			width: PropTypes.number.isRequired,
			height: PropTypes.number.isRequired,
		} ),

		style: PropTypes.object,

		/**
		 * zoom (required if markers not present) defines the zoom level of the map,
		 * which determines the magnification level of the map.
		 *
		 * @see https://developers.google.com/maps/documentation/staticmaps/intro#Zoomlevels
		 */
		zoom: PropTypes.number.isRequired,

		/**
		 * scale affects the number of pixels that are returned.
		 * scale=2 returns twice as many pixels as scale=1 while retaining the same coverage area and level of detail
		 * The default value is calculated from the screen PixelRatio.
		 */
		scale: PropTypes.number,

		/**
		 * @see https://developers.google.com/maps/documentation/staticmaps/intro#ImageFormats
		 */
		format: PropTypes.oneOf( IMAGE_FORMATS_VALUES ),

		/**
		 * @see https://developers.google.com/maps/documentation/staticmaps/intro#MapTypes
		 */
		mapType: PropTypes.oneOf( MAP_TYPES_VALUES ),

		/**
		 * Add a marker on the center
		 */
		hasCenterMarker: PropTypes.bool,
		apiKey: PropTypes.string,
	}

	static defaultProps = {
		format: IMAGE_FORMATS.JPG,
		mapType: MAP_TYPES.ROADMAP,
		hasCenterMarker: true,
		style: {},
		scale: 2,
		interactive: false,
		apiKey: API_KEY,
		zoom: DEFAULT_ZOOM,
		coordinates: {},
		address: {},
	}

	constructor( props ) {
		super( ...arguments );

		this.state = {
			...props,
			error: '',
			isLoading: true,
			rendered: false,
		};

		this.interactiveMapContainer = React.createRef();
		this.map = {
			instance: null,
			marker: null,
		};
		this.interval = noop;
		this.tries = 0;
		this.MAX_TRIES = 5;
	}

	/**
	 * @todo  We need to not do this Logic over in this template
	 * @see   https://github.com/moderntribe/events-gutenberg/pull/327#discussion_r219090823
	 */

	static getDerivedStateFromProps( nextProps, prevState ) {
		// return if it was rendered already
		if ( true === prevState.rendered ) {
			return null;
		}

		// return if the coordinates haven't changed
		if (
			nextProps.coordinates.lat === prevState.coordinates.lat &&
			nextProps.coordinates.lng === prevState.coordinates.lng
		) {
			return null;
		}

		return { rendered: true, loadingMap: false };
	}

	componentDidMount() {
		this.loadMap();
	}

	loadMap() {
		if ( ! google() ) {
			this.tryAgain();
			return;
		}

		const { maps } = google();
		// Try to fetch the library 0.5 seconds later
		if ( ! maps ) {
			this.tryAgain();
			return;
		}

		// There's no valid coordinates, fallback to the image map.
		if ( this.invalidLocation() ) {
			const { address } = this.props;

			if ( isEmpty( address ) ) {
				this.setState( {
					interactive: false,
					isLoading: false,
					error: __(
						'The map does not have valid coordinates nor a valid address',
						'the-events-calendar',
					),
				} );
				return;
			}

			this.setState( {
				interactive: false,
				isLoading: false,
			} );
			return;
		}

		this.setState( {
			isLoading: false,
			interactive: true,
		}, this.attachInteractiveMap );
	}

	getMapConfig() {
		const {
			zoom,
			mapType,
		} = this.props;

		const type = isArray( mapType ) ? mapType : [ mapType ];

		return {
			center: this.getLocation(),
			zoom: zoom,
			mapTypeControl: type.length > 1,
			mapTypeControlOptions: {
				mapTypeIds: type,
			},
			streetViewControl: false,
			fullscreenControl: false,
		};
	}

	invalidLocation() {
		const location = this.getLocation();
		const { lat, lng } = location;
		return ! lat || ! lng;
	}

	getLocation() {
		const { coordinates } = this.props;
		const { lat, lng } = coordinates;

		return {
			lat,
			lng,
		};
	}

	tryAgain = () => {
		if ( this.interval ) {
			clearInterval( this.interval );
		}

		if ( this.tries >= this.MAX_TRIES ) {
			this.setState( {
				isLoading: false,
				error: __(
					'Make sure Google Maps Library is included on this page.',
					'the-events-calendar',
				),
			} );
			return;
		}

		this.interval = setInterval( () => {
			this.loadMap();
		}, 500 );

		this.tries += 1;
	}

	render() {
		const { isLoading, rendered, loadingMap } = this.state;
		const containerClass = classNames( 'tribe-editor__map', {
			'tribe-editor__map--loading': isLoading,
		} );

		let renderMap = this.renderMap();

		if (
			true === rendered &&
			true !== loadingMap
		) {
			renderMap = this.renderMapUpdate();
		}

		return (
			<div className={ containerClass }>
				{ renderMap }
			</div>
		);
	}

	renderMapUpdate() {
		this.setState( { loadingMap: true } );
		this.loadMap();
		return this.renderMap();
	}

	renderMap() {
		const { isLoading, error, interactive, apiKey, rendered } = this.state;

		if ( isLoading ) {
			return <Spinner />;
		}

		if ( error ) {
			return ( <h4>{ error }</h4> );
		}

		if ( ! apiKey ) {
			return (
				<h4>
					{ __( 'A Google Map API KEY is required to view the map', 'the-events-calendar' ) }
				</h4>
			);
		}

		if ( rendered || interactive ) {
			return this.renderInteractive();
		}

		return this.renderIframe();
	}

	renderImage() {
		return (
			<picture className="tribe-editor__map--static">
				<img
					className="tribe-element-map-object"
					alt="map"
					src={ this.mapUrl }
				/>
				<div className="trie-editor__spinner__container">
					<Spinner />
				</div>
			</picture>
		);
	}

	renderInteractive() {
		return (
			<section className="tribe-editor__map--interactive">
				<div className="tribe-editor__map--dynamic" ref={ this.interactiveMapContainer }>
				</div>
				<div className="trie-editor__spinner__container">
					<Spinner />
				</div>
			</section>
		);
	}

	renderIframe() {
		return (
			<iframe
				title="Venue Map"
				src="https://www.google.com/maps/embed?pb=!1m10!1m8!1m3!1d65369183.36050215!2d0!3d0!3m2!1i1024!2i768!4f13.1!5e0!3m2!1sen!2sus!4v1633005420084!5m2!1sen!2sus" // eslint-disable-line max-len
				width="450px"
				height="300px"
			/>
		);
	}

	attachInteractiveMap = () => {
		const { interactive } = this.state;
		const { interactiveMapContainer, map } = this;

		if ( ! interactive || ! interactiveMapContainer.current ) {
			return this.renderImage();
		}

		const { maps } = google();

		map.instance = new maps.Map(
			interactiveMapContainer.current,
			this.getMapConfig(),
		);

		if ( map.instance ) {
			map.marker = new maps.Marker( {
				position: this.getLocation(),
				map: map.instance,
			} );
		}
	}

	get mapUrl() {
		const {
			zoom,
			size,
			scale,
			format,
			mapType,
			apiKey,
			address,
		} = this.props;

		const { width, height } = size;

		const queryArgs = {
			zoom: zoom,
			maptype: mapType,
			key: apiKey,
		};
		let rootUrl = null;
		const { interactive } = this.state;

		const coordinates = this.getLocation();
		const { lat, lng } = coordinates;
		if ( interactive ) {
			rootUrl = this.constructor.RootEmbedUrl;

			queryArgs.q = `${ lat },${ lng }`;
		} else {
			rootUrl = this.constructor.RootStaticUrl;

			queryArgs.scale = scale;
			queryArgs.size = `${ width }x${ height }`;
			queryArgs.format = format;

			const invalid = this.invalidLocation();
			if ( invalid && ! isEmpty( address ) ) {
				queryArgs.center = address;
			} else {
				queryArgs.center = `${ lat },${ lng }`;
			}
			queryArgs.markers = this.markerParams;
		}

		return `${ rootUrl }?${ stringify( queryArgs ) }`;
	}

	get markerParams() {
		const { hasCenterMarker, address } = this.props;

		const coordinates = this.getLocation();
		const { lat, lng } = coordinates;

		const invalid = this.invalidLocation();
		const marker = invalid ? address : `${ lat },${ lng }`;
		const markerParams = `size:mid|color:0xff0000|label:|${ marker }`;
		return hasCenterMarker ? markerParams : '';
	}
}
