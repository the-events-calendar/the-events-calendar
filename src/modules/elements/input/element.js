/**
 * External dependencies
 */
import { isFunction, noop } from 'lodash';
import PropTypes from 'prop-types';
import validator from 'validator';

/**
 * WordPress dependencies
 */
import { Component } from '@wordpress/element';
import { TextControl } from '@wordpress/components';

/**
 * Internal dependencies
 */
import './style.pcss';

/**
 * Module Code
 */

/**
 * Input element that adds two important properties:
 *
 * - required
 * - validate
 *
 * Both properties are helpful to provide validation based on the type of the element also has
 * the option to add a cusotm validation callback with the property validateCallback.
 *
 * <Input />
 *
 * Abstraction on top of the native <input> element it gives external instance methods such as:
 *
 * - isValid()
 * - focus()
 */
class Input extends Component {
	/**
	 * Default types for properties required for this component
	 *
	 * @param validated {boolean} If this component needs to be validated
	 * @param required {boolean} If this input is required
	 *
	 * @type {{validated: shim, required: shim}}
	 */
	static propTypes = {
		validate: PropTypes.bool,
		required: PropTypes.bool,
	}

	/**
	 * Set the default values for the required properties if not provided
	 *
	 * @type {{required: boolean, validate: boolean}}
	 */
	static defaultProps = {
		required: false,
		validate: false,
	}

	constructor() {
		super( ...arguments );

		this.state = {
			isValid: this.validate( '' ),
		};
	}

	/**
	 * Callback fired every time the input changes, if the property onChange is passed to the component is called as well
	 * synchronously.
	 *
	 * @param {Event} event Event fired by the onChange listener
	 */
	onChange = ( input ) => {
		const { onChange, onComplete, validate } = this.props;
		const callback = isFunction( onChange ) ? onChange : noop;
		const completeCallback = isFunction( onComplete ) ? onComplete : noop;

		if ( validate ) {
			this.setState( () => ({ isValid: this.validate( input ) }), completeCallback );
			callback( input );
		} else {
			completeCallback();
			callback( input );
		}
	}

	/**
	 * Validates the component using validateCallback if provided or using the logic based on the type inferring
	 *
	 * @param {string} value The value to be validated
	 * @returns {boolean} `true` if the param is valid false otherwise
	 */
	validate( value ) {
		const { validateCallback } = this.props;
		return isFunction( validateCallback ) ? validateCallback( value ) : this.maybeValidate( value );
	}

	/**
	 * If the input is empty does not make any validation unless the value is required otherwise it uses the validation
	 * based on the type="url" of the <input> component.
	 *
	 * @param {string} value The value to be validated
	 * @returns {boolean} `true` if the param is valid false otherwise
	 */
	maybeValidate = ( value ) => {
		const { type, required } = this.props;
		
		if ( value.length === 0 ) {
			return ! required;
		}
		
		let isValid = true;
		switch ( type ) {
			case 'tel':
			case 'phone':
				isValid = validator.isMobilePhone( value, 'any' );
				break;
			case 'email':
				isValid = validator.isEmail( value );
				break;
			case 'url':
				isValid = validator.isURL( value );
				break;
			case 'number':
				isValid = validator.isNumeric( value );
				break;
		}
		return isValid;
	}

	/**
	 * If the component is valid or not based on the validation logic
	 *
	 * @returns {boolean}
	 */
	isValid() {
		return this.state.isValid;
	}

	/**
	 * Focus to the current component input
	 */
	focus() {
		this.input.focus();
	}

	/**
	 * If the Component is valid is going to Append the class: `is-valid` otherwise if it fails is going to use the
	 * class `is-invalid`.
	 *
	 * @returns {string|*}
	 */
	getClassName() {
		const { className, validate } = this.props;
		const { isValid } = this.state;
		const classes = className ? className.split( ' ' ) : [];

		if ( validate ) {
			if ( isValid ) {
				classes.push( 'tribe-editor--valid' );
			} else {
				classes.push( 'tribe-editor--valid' );
			}
		}

		return classes
			.filter( ( name ) => name && name.length )
			.join( ' ' );
	}

	render() {
		// Remove properties that are not part of the DOM.
		const { onComplete, required, validate, ...properties } = this.props;
		return (
			<TextControl
				{ ...properties }
				className={ `${ this.getClassName() }` }
				ref={ ( input ) => this.input = input }
				onChange={ this.onChange }
			/>
		);
	}
}

export default Input;
