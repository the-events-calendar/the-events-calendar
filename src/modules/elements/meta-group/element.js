/**
 * External dependencies
 */
import React from 'react';
import classNames from 'classnames';
import { PropTypes } from 'prop-types';

/**
 * WordPress dependencies
 */
import { Component } from '@wordpress/element';

/**
 * Internal dependencies
 */

/**
 * Module Code
 */
class MetaGroup extends Component {
	static propTypes = {
		className: PropTypes.string,
		children: PropTypes.node,
		groupKey: PropTypes.string,
	};

	static defaultProps = {
		className: '',
		children: null,
		groupKey: '',
	};

	render() {
		const { groupKey, className, children } = this.props;

		const names = classNames( [
			'tribe-editor__meta-group',
			`tribe-editor__meta-group--${ groupKey }`,
			className,
		] );
		return (
			<div
				className={ names }
				key={ groupKey }
			>
				{ children }
			</div>
		);
	}
}

export default MetaGroup;
