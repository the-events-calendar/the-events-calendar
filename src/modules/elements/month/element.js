/**
 * External dependencies
 */
import React from 'react';
import { omit, noop } from 'lodash';
import DayPicker, { DateUtils } from 'react-day-picker';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import moment from 'moment';

/**
 * Wordpress dependencies
 */
import { Component } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { YearMonthForm } from '@moderntribe/events/elements';
import './style.pcss';

const today = new Date();
const currentYear = today.getFullYear();
const currentMonth = today.getMonth();
const yearsBack = 5;
const fromMonth = new Date( currentYear - yearsBack, currentMonth );
const toMonth = new Date( currentYear + 10, 11 );

export default class Month extends Component {
	static propTypes = {
		withRange: PropTypes.bool,
		onSelectDay: PropTypes.func,
		from: PropTypes.instanceOf( Date ),
		to: PropTypes.instanceOf( Date ),
		month: PropTypes.instanceOf( Date ),
		setVisibleMonth: PropTypes.func,
	};

	static defaultProps = {
		onSelectDay: noop,
		from: today,
		to: undefined,
		month: fromMonth,
		setVisibleMonth: noop,
	};

	constructor() {
		super( ...arguments );

		this.state = {
			toMonth: toMonth,
			from: null,
			to: null,
		};
	}

	selectDay = ( day ) => {
		const { withRange } = this.props;
		let range = {};

		if ( withRange ) {
			range = DateUtils.addDayToRange( day, this.state );

			// if the range was unselected we fallback to the first available day
			if ( range.from === null && range.to === null ) {
				range.from = today;
				range.to = undefined;
			}

			if ( range.to && moment( range.to ).isSame( range.from ) ) {
				range.to = undefined;
			}
		} else {
			range.from = day;
			range.to = undefined;
		}

		this.setState( this.maybeUpdate( range ), () => {
			this.onSelectCallback();
		} );
	};

	maybeUpdate = ( range ) => ( state ) => {
		if ( state.from === range.from && state.to === range.to ) {
			return null;
		}
		return range;
	};

	onSelectCallback = () => {
		const { onSelectDay } = this.props;
		onSelectDay( omit( this.state, [ 'withRange' ] ) );
	};

	getSelectedDays = () => {
		const { withRange, from, to } = this.props;
		if ( withRange ) {
			return [ from, { from, to } ];
		}
		return from;
	};

	getCaptionElement = ( { date, localeUtils } ) => {
		const { month, setVisibleMonth } = this.props;

		if ( date.getMonth() !== month.getMonth() ) {
			return this.renderCaption( date, localeUtils );
		}

		return (
			<YearMonthForm
				today={ today }
				date={ date }
				localeUtils={ localeUtils }
				onChange={ setVisibleMonth }
			/>
		);
	};

	renderCaption = ( date, localeUtils ) => (
		<div className={ 'tribe-editor__daypicker-caption' }>
			<div>
				{ localeUtils.formatMonthTitle( date ) }
			</div>
		</div>
	);

	render() {
		const { from, to, month, withRange, setVisibleMonth } = this.props;
		const modifiers = withRange ? { start: from, end: to } : {};
		const containerClass = classNames( { 'tribe-editor__calendars--range': withRange } );
		return (
			<DayPicker
				className={ containerClass }
				fromMonth={ fromMonth }
				toMonth={ this.state.toMonth }
				month={ month }
				numberOfMonths={ 2 }
				modifiers={ modifiers }
				selectedDays={ this.getSelectedDays() }
				onDayClick={ this.selectDay }
				onMonthChange={ setVisibleMonth }
				captionElement={ this.getCaptionElement }
			/>
		);
	}
}
