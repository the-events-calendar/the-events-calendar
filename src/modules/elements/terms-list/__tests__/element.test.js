/**
 * External dependencies
 */
import React from 'react';
import renderer from 'react-test-renderer';

/**
 * Internal dependencies
 */
import { TaxonomiesElement } from '../element';

describe( 'Taxonomies Element', () => {
	it( 'renders empty when items is empty', () => {
		const tree = renderer.create( <TaxonomiesElement terms={ [] } /> );
		expect( tree.toJSON() ).toEqual( null );
	} );

	it( 'renders Empty property', () => {
		const tree = renderer.create( <TaxonomiesElement terms={ [] } renderEmpty={ 'Hi!' } /> );
		expect( tree.toJSON() ).toMatchSnapshot();
	} );

	it( 'renders spinner when requesting', () => {
		const tree = renderer.create(
			<TaxonomiesElement terms={ [] } isRequesting={ true } />,
		);
		expect( tree.toJSON() ).toMatchSnapshot();
	} );

	it( 'Should render the list of items', () => {
		const items = [
			{
				description: '',
				id: 1,
				meta: [],
				name: 'modern',
				slug: 'modern',
				taxonomy: 'post_tag',
			},
			{
				description: '',
				id: 2,
				meta: [],
				name: 'tribe',
				slug: 'tribe',
				taxonomy: 'post_tag',
			},
		];
		const tree = renderer.create(
			<TaxonomiesElement terms={ items } isRequesting={ false } />,
		);
		expect( tree.toJSON() ).toMatchSnapshot();
	} );
} );
