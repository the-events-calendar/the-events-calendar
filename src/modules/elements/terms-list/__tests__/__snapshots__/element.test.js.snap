// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Taxonomies Element Should render the list of items 1`] = `
<div
  className="tribe-editor__terms "
>
  <strong
    className="tribe-editor__terms__label"
  >
     
  </strong>
  <div
    className="tribe-editor__terms__list-wrapper"
  >
    <ul
      className="tribe-editor__terms__list tribe-editor__terms__list--level-0"
    >
      <li
        className="tribe-editor__terms__list-item tribe-editor__terms__list-item--level-0"
      >
        <a
          className="tribe-editor__terms__list-item-link"
          href="/events/tag/modern"
          rel="noopener noreferrer"
          target="_blank"
        >
          modern
        </a>
        <span>
          , 
        </span>
      </li>
      <li
        className="tribe-editor__terms__list-item tribe-editor__terms__list-item--level-0"
      >
        <a
          className="tribe-editor__terms__list-item-link"
          href="/events/tag/tribe"
          rel="noopener noreferrer"
          target="_blank"
        >
          tribe
        </a>
        
      </li>
    </ul>
  </div>
</div>
`;

exports[`Taxonomies Element renders Empty property 1`] = `
<div
  className="tribe-editor__terms--empty"
>
  <strong
    className="tribe-editor__terms__label"
  >
     
  </strong>
  Hi!
</div>
`;

exports[`Taxonomies Element renders spinner when requesting 1`] = `
<div
  className="tribe-editor__terms "
>
  <strong
    className="tribe-editor__terms__label"
  >
     
  </strong>
  <div
    className="tribe-editor__terms__list-wrapper"
  >
    <div
      className="tribe-editor__terms__spinner "
    >
      <strong
        className="tribe-editor__terms__label"
      >
         
      </strong>
      üèÉ‚Äç‚ôÇÔ∏è
    </div>
  </div>
</div>
`;
